plugins/wasi_nn/wasinn_whisper.cpp

```diff
@@ -16,6 +18,309 @@ namespace WasmEdge::Host::WASINN::Whisper {
 #ifdef WASMEDGE_PLUGIN_WASI_NN_BACKEND_WHISPER
 
 namespace {
+int timestampToSample(int64_t T, int NSamples, int WhisperSampleRate) {
+  return std::max(0, std::min(static_cast<int>(NSamples) - 1,
+                              static_cast<int>((T * WhisperSampleRate) / 100)));
+}
+std::string toTimestamp(int64_t T, bool Comma) {
+  int64_t Msec = T * 10;
+  int64_t Hr = Msec / (1000 * 60 * 60);
+  Msec = Msec - Hr * (1000 * 60 * 60);
+  int64_t Min = Msec / (1000 * 60);
+  Msec = Msec - Min * (1000 * 60);
+  int64_t Sec = Msec / 1000;
+  Msec = Msec - Sec * 1000;
+
+  char Buf[32];
+  snprintf(Buf, sizeof(Buf), "%02d:%02d:%02d%s%03d", static_cast<int>(Hr),
+           static_cast<int>(Min), static_cast<int>(Sec), Comma ? "," : ".",
+           static_cast<int>(Msec));
+
+  return std::string(Buf);
+}
+
+std::string
+estimateDiarizationSpeaker(const std::vector<std::vector<float>> PCMF32s,
+                           int64_t T0, int64_t T1, bool IdOnly = false) {
+  std::string Speaker = "";
+  const int64_t NSamples = PCMF32s[0].size();
+
+  const int64_t is0 = timestampToSample(T0, NSamples, WHISPER_SAMPLE_RATE);
+  const int64_t is1 = timestampToSample(T1, NSamples, WHISPER_SAMPLE_RATE);
+
+  double energy0 = 0.0f;
+  double energy1 = 0.0f;
+
+  for (int64_t j = is0; j < is1; j++) {
+    energy0 += fabs(PCMF32s[0][j]);
+    energy1 += fabs(PCMF32s[1][j]);
+  }
+
+  if (energy0 > 1.1 * energy1) {
+    Speaker = "0";
+  } else if (energy1 > 1.1 * energy0) {
+    Speaker = "1";
+  } else {
+    Speaker = "?";
+  }
+
+  if (!IdOnly) {
+    Speaker.insert(0, "(speaker ");
+    Speaker.append(")");
+  }
+
+  return Speaker;
+}
+
+bool outputSrt(whisper_context *Ctx, const std::string &Fname,
+               const Config &Params,
+               const std::vector<std::vector<float>> &PCMF32s) {
+  std::ofstream Fout(Fname);
+  if (!Fout.is_open()) {
+    spdlog::error("[WASI-NN] Whisper backend: failed to open {} for writing."sv,
+                  Fname);
+    return false;
+  }
+  spdlog::info("[WASI-NN] Whisper backend: saving srt output to {}."sv, Fname);
+
+  const int NSegments = whisper_full_n_segments(Ctx);
+  for (int I = 0; I < NSegments; ++I) {
+    const std::string &Text = whisper_full_get_segment_text(Ctx, I);
+    const int64_t T0 = whisper_full_get_segment_t0(Ctx, I);
+    const int64_t T1 = whisper_full_get_segment_t1(Ctx, I);
+    std::string Speaker = "";
+
+    if (Params.Diarize && PCMF32s.size() == 2) {
+      Speaker = estimateDiarizationSpeaker(PCMF32s, T0, T1);
+    }
+
+    Fout << I + 1 + Params.OffsetN << "\n";
+    Fout << toTimestamp(T0, true) << " --> " << toTimestamp(T1, true) << "\n";
```

Comment:
Also, I have a question: Does whisper.cpp provide an API for enabling SRT output? If so, could we use it instead of handling the output inside the plugin?

Reply:
I guess whisper.cpp does not have this API. Because its [example](https://github.com/ggerganov/whisper.cpp/blob/7d55637f0bc8ff1b405f6f18c5c04fe39e4ad389/examples/cli/cli.cpp#L427) also implements it again.